### DISCUSSION TITLE
How to cancel async job that already running?

### DISCUSSION THREAD

**Discussion by Aldiwildan77** [go-co-op/gocron#204] [Q&A] [answered]
Hello, I have got a project to run a scheduler using this package. But I got confused about the scheduler cancellation, is there any example of how to use that? 

*this is my example code. I wanna cancel if the task inside this job got an error so the job itself will be cancelled.

```go
func main() {
    loc, _ := time.LoadLocation(utils.TIMEZONE_LOCAL)
    s := gocron.NewScheduler(loc)
    s.StartAsync()
    s.Every(10).Second().Do(RecoverTransactionScheduler)
}

func RecoverTransactionScheduler() {
    // run task
    // if error occur here, stop this job itself
}
```

**Comment by JohnRoesler** [comment, marked as answer]
Hi @Aldiwildan77 

Yes - you can stop the job by removing it. https://pkg.go.dev/github.com/go-co-op/gocron#Scheduler.Remove

As noted, you could also add a context to the job function to allow stopping the job if it's long running and you've removed it.

**Reply by Aldiwildan77** [reply]
is the remove method can be used inside the scheduler task itself?

**Reply by Aldiwildan77** [reply]
sorry but is there any example for context usage?

**Reply by JohnRoesler** [reply]
> is the remove method can be used inside the scheduler task itself?

Yes, this on the scheduler to remove the job itself. It will stop the job from making any additional runs.

> sorry but is there any example for context usage?

To stop a job run yourself without removing it from the scheduler, you can pass a context with cancellation into the job func and determine how you want to handle a cancel event. For example:

```golang
package main

import (
	"context"
	"log"
	"time"

	"github.com/go-co-op/gocron"
)

func main() {
	s := gocron.NewScheduler(time.Local)

	ctx, cancel := context.WithCancel(context.Background())

	s.Every(5).Seconds().Do(func(ctx context.Context) {
		for {
			select {
			case <-ctx.Done():
				return
			default:
			}
			log.Println("job started")
			select {
			case <-ctx.Done():
				return
			default:
			}
			time.Sleep(5 * time.Second)
			select {
			case <-ctx.Done():
				return
			default:
			}
			log.Println("job done")
			return
		}

	}, ctx)

	s.StartAsync()

	time.Sleep(1 * time.Second)

	log.Println("canceling job")
	cancel()
}

```
